### DISCUSSION TITLE
High Availability

### DISCUSSION THREAD

**Discussion by irtaza9** [Dokploy/dokploy#748] [Ideas]
I want to achieve the HA for my services. Is there any way to achieve this by using dokploy?

**Comment by Siumauricio** [comment]
Hi, yes you can achieve HA using dokploy, however you need a external load balancer, probably in the next weeks i will make a guide, people on discord asked the same, but yes is possible

**Reply by m7mdcc** [reply]
i think the issue will be syncing data within the nodes , like app file changes , and database .. ( hint : shared volumes )

maybe rclone may help here

**Reply by Siumauricio** [reply]
I'll suggest to use stateless applications instead of trying to achieve HA with databases...

**Reply by irtaza9** [reply]
we need a guide on HA for applications and databases which will definitely help everyone, Also it will add more value to this amazing product. It is not a big deal to add external lb/HAProxy but your suggested way and guide will make a priority.

**Reply by Siumauricio** [reply]
For databases I have never tried it, but surely the guide I make will be for applications only.

**Reply by irtaza9** [reply]
@Siumauricio any update on this?

**Reply by ravinataraju** [reply]
I have currently migrated to cloudflare dns and have setup load balancer. I have setup an application (3 replicas) pointing to test domain which is load balanced by cloudflare, one ip mapped to dokploy installed vps for which health monitor is ok, but the ip of the secondary and third master health monitor is not working. The idea is for any reason the first master is down, docker swarm will take over on either secondary or third master so the internet traffic will still get a response. I might be wrong and looking forward to the HA documentation to do it properly. I kindly request to make this a high priority

**Reply by kaveh-ahmadian** [reply]
@Siumauricio was the guide ever published?

**Reply by jameshewitt1399** [reply]
@Siumauricio any update?

**Reply by jan-paulus** [reply]
@Siumauricio any update on this? Would be great if this was documented more.

**Comment by ravinataraju** [comment]
It would be beneficial to add the usage of swarm and remote server in HA, also a check box to setup replica database whilst setting up the database either within the swarm or remote servers whichever is the correct way to do it

**Comment by sunnycatnice** [comment]
Implementing a complete HA (High Availability) solution, as mentioned before, including shared storage across machines—even for the database, not just the application—would be ideal. Anyway I encountered issues because setting up HA with Docker Swarm for my volumes is quite challenging, and there doesn’t seem to be a "standardized" approach for achieving this.

**Reply by ravinataraju** [reply]
i found this solution, haven't tested as i have a mysql container running in one of the node, you can try https://github.com/jnidzwetzki/mysql-ha-cloud/blob/main/docs/deployment-docker-swarm.md

**Reply by ravinataraju** [reply]
Hi there, I was wondering if there was any progress on high availability options, as that is probably the most missing feature in dokploy using multiple server and swarm support

**Reply by Siumauricio** [reply]
Nothing restricts you to have HA, I currently have HA for dokploy cloud and it works fine, just implement an external load balancer and that would be enough

**Reply by voale-rogii** [reply]
Hello team, you tell us about HA, but yours [script](https://docs.dokploy.com/docs/core/manual-installation#installation-script) installs traefik only one node. 
Lets suppose there is a haproxy to node1, node2, node3 with healthcheks and there is an A dns record to haproxy host.
I'm installing dokploy on node1 (traefik with rules installed only here)  and then adding 2 master nodes in swarm mode.
Then suppose node1 is making unavailable. 
Describe us please where is HA here?

**Reply by BobbyMoore** [reply]
@voale-rogii - as far as I understand you would need Traefik on an external/separate host

**Reply by 4volodin** [reply]
"Traefik on an external/separate host" - it's not the best case because it's need setup a tcp access to the docker on master nodes(for traefik magic)
I think the best case is:
- self install docker swarm with min 3 master nodes
- self install traefik
with
```
--providers.docker=true
--providers.docker.swarmMode=true
```
..
```
deploy:
      mode: global
      placement:
        constraints:
          - node.role == manager
```
- tls termination on external host with vrrp or using for example AWS ALB
sure,  you can terminate tls on the traefik + consul for storing certs without external host

**Reply by 4volodin** [reply]
but I don't know how will dokploy work with 3 traefik replicas ?

consul https://doc.traefik.io/traefik/v1.7/user-guide/cluster/
